### PR TITLE
fix kubelet image gc

### LINKED ISSUE #119642: [kubelet-image-gc] minimum-image-ttl-duration not work,when kubelet restart.
### What happened?

when kubelet restart,all images which notused cleaned.
but there is a flag in kubelet,`minimum-image-ttl-duration`,explain：Minimum age for an unused image before it is garbage collected.  Examples: '300ms', '10s' or '2h45m'.
i set `minimum-image-ttl-duration` to `20m0s`,it's means if a image not used,it will live 20m and then clean by kubelet（if has disk pressure）.if kubelet not restart, it work,but if restart ther is a bug, `minimum-image-ttl-duration` not work,image will be clean，even if it's no enough  time.

### What did you expect to happen?

`minimum-image-ttl-duration` work, image even if not used,it will live 20m and then clean by kubelet.

### How can we reproduce it (as minimally and precisely as possible)?

1）make node trigger disk pressure.
2）pull some image which not used by pod
3）restart kubelet
4）image cleaned by kubelet after around 10s

### Anything else we need to know?

i see the source code, and maybe there is a bug.

in [detectImages](https://github.com/kubernetes/kubernetes/blob/97c7dbcd227ec253bb277a91575706fe30f519dc/pkg/kubelet/images/image_gc_manager.go#L261-L262) func,well detect image and update firstdetect time for every image.
```go
	// New image, set it as detected now.
		if _, ok := im.imageRecords[image.ID]; !ok {
			klog.V(5).InfoS("Image ID is new", "imageID", image.ID)
			im.imageRecords[image.ID] = &imageRecord{
				firstDetected: detectTime,
			}
		}
```
and in  [freeSpace](https://github.com/kubernetes/kubernetes/blob/97c7dbcd227ec253bb277a91575706fe30f519dc/pkg/kubelet/images/image_gc_manager.go#L394C1-L397) func,well check images livetime by firstDetected.
```go
		if freeTime.Sub(image.firstDetected) < im.policy.MinAge {
			klog.V(5).InfoS("Image ID's age is less than the policy's minAge, not eligible for garbage collection", "imageID", image.id, "age", freeTime.Sub(image.firstDetected), "minAge", im.policy.MinAge)
			continue
		}
```
if kubelet running normal,it work,because there is a [Start](https://github.com/kubernetes/kubernetes/blob/97c7dbcd227ec253bb277a91575706fe30f519dc/pkg/kubelet/images/image_gc_manager.go#L190-L204) func.
```go
func (im *realImageGCManager) Start() {
	ctx := context.Background()
	go wait.Until(func() {
		// Initial detection make detected time "unknown" in the past.
		var ts time.Time
		if im.initialized {
			ts = time.Now()
		}
                // in first time,ts is zero. 
               // in first time,ts is zero.
		_, err := im.detectImages(ctx, ts)
		if err != nil {
			klog.InfoS("Failed to monitor images", "err", err)
		} else {
			im.initialized = true
		}
	}, 5*time.Minute, wait.NeverStop)
//...

```
in this line,`_, err := im.detectImages(ctx, ts)`,in first time,ts is zero. so image's firstDetected time is zero,and in freeSpace
`reeTime.Sub(image.firstDetected) < im.policy.MinAge` this check is always false,so image will be cleaned.

maybe we can change to `im.detectImages(time.now())`,can fix this bug.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.6", GitCommit:"ad3338546da947756e8a88aa6822e9c11e7eac22", GitTreeState:"clean", BuildDate:"2022-04-14T08:49:13Z", GoVersion:"go1.17.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.6", GitCommit:"ad3338546da947756e8a88aa6822e9c11e7eac22", GitTreeState:"clean", BuildDate:"2022-04-14T08:43:11Z", GoVersion:"go1.17.9", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
None
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
$ uname -a
Linux kc-worker2 5.4.156-1.el7.elrepo.x86_64 #1 SMP Tue Oct 26 11:19:09 EDT 2021 x86_64 x86_64 x86_64 GNU/Linux

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by lixd**
#### What type of PR is this?
/kind bug

#### Which issue(s) this PR fixes:

Fixes #119642

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by ffromani** [comment]
Another option could be to store the timestamp of the last check on a checkpoint file (/var/lib/kubelet/...).
In normal operation this checkpoint would be ignored,because kubelet can keep the necessary state in memory, but on kubelet restart/node reboot this will give us a way to do a precise recovery.
Additionally, if the checkpoint is missing, kubelet can happily keep its current behavior (either the one fixed in this PR or the previous).
I'm not sure the extra complexity is worht the gains, and I'm not approver so I can't really say, but I for myself would explore this way.

File: pkg/kubelet/images/image_gc_manager.go [hunk 1]
```diff
@@ -190,16 +186,9 @@ func NewImageGCManager(runtime container.Runtime, statsProvider StatsProvider, r
 func (im *realImageGCManager) Start() {
 	ctx := context.Background()
 	go wait.Until(func() {
-		// Initial detection make detected time "unknown" in the past.
-		var ts time.Time
-		if im.initialized {
-			ts = time.Now()
-		}
-		_, err := im.detectImages(ctx, ts)
+		_, err := im.detectImages(ctx, time.Now())
```
Comment (ffromani) [inline pkg/kubelet/images/image_gc_manager.go:189]: thinking about it, previously a kubelet restart triggered a GC (which is exactly what is being fixed here) and I wonder if there are ppl out there depending on this to free disk space on nodes with high disk pressure.
However now we need to wait a bit more (basically the actual min age to expire after a restart) so it should not be a big deal even for these folks.

**Review by ffromani — COMMENTED** [review]
/lgtm

I think that adding a state file like I proposed can be a further enhancement, if we ever need to get there.

**Comment by ffromani** [comment]
let's hear thoughts from other reviewers and approvers though

**Comment by bart0sh** [comment]
@lixd Since this is a bugfix, can we have this fix covered by the unit or/and e2e tests?

**Comment by bart0sh** [comment]
@lixd Can you explain what kind of race did you fix with your last commit in this PR?

**Comment by lixd** [comment]
> @lixd Can you explain what kind of race did you fix with your last commit in this PR?

the fakeruntime'  ListImage func return a slice,and the caller(manager) got this slice and will wirte it by `im.imageCache.set(images)`, finally func which write it is `sort.sort`,the Swap func `func (a ByImageSize) Swap(i, j int) { a[i], a[j] = a[j], a[i] }`,and the fakeruntime's  `GetImageRef` will try to read the slice, so there is a  data race, i use snapshot func to solve it.

**Comment by bart0sh** [comment]
@lixd Would it be better/more logical to use locking instead of snapshotting?

**Comment by lixd** [comment]
@bart0sh i think runtime do not need to share variable with image manager,so maybe snapshotting it better.

**Comment by bart0sh** [comment]
@lixd 
> i think runtime do not need to share variable with image manager,so maybe snapshotting it better.

I'm still not convinced. There is a race between `ImageGCmanager` and `imageCache` operations and it needs to be fixed  in a generic manner, e.g. by introducing one lock that's used by both components. Currently [cache operation is locked](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/images/image_gc_manager.go#L128), but manager doesn't use that lock.

Let's see what other reviewers say.
@haircommander @ffromani PTAL: https://github.com/kubernetes/kubernetes/pull/119652#issuecomment-1693077710
/cc @haircommander

**Comment by lixd** [comment]
i am confused，FakeRuntime'ListImages  just return a  private variable,`ImageList`
```go
type FakeRuntime struct {
	sync.Mutex
	ImageList         []kubecontainer.Image
}
func (f *FakeRuntime) ListImages(_ context.Context) ([]kubecontainer.Image, error) {
	f.Lock()
	defer f.Unlock()

	f.CalledFunctions = append(f.CalledFunctions, "ListImages")
	return f.ImageList, f.Err
}
```
and the `ImageGCManager` call ListImages get this private variable and write to it,
```go
func (im *realImageGCManager) Start() {
	ctx := context.Background()
	go wait.Until(func() {
		_, err := im.detectImages(ctx, time.Now())
		if err != nil {
			klog.InfoS("Failed to monitor images", "err", err)
		}
	}, 5*time.Minute, wait.NeverStop)

	// Start a goroutine periodically updates image cache.
	go wait.Until(func() {
		images, err := im.runtime.ListImages(ctx)
		if err != nil {
			klog.InfoS("Failed to update image list", "err", err)
		} else {
			im.imageCache.set(images)
		}
	}, 30*time.Second, wait.NeverStop)

}

func (i *imageCache) set(images []container.Image) {
	i.Lock()
	defer i.Unlock()
	// The image list needs to be sorted when it gets read and used in
	// setNodeStatusImages. We sort the list on write instead of on read,
	// because the image cache is more often read than written
	sort.Sort(sliceutils.ByImageSize(images))   // there will write to images
	i.images = images
}
```
core is `sort.Sort(sliceutils.ByImageSize(images))   // there will write to images`

so the race is between `FakeRuntime` and `ImageGCManager`,instead of `ImageGCManager` and `imageCache`，Maybe I didn't make it clear, please take a look.
@bart0sh

**Comment by bart0sh** [comment]
These 2 statements look mutually exclusive to me as sort.Sort is called from **imageCache**.set

> so the race is between FakeRuntime and ImageGCManager,instead of ImageGCManager and imageCache

...

> core is sort.Sort(sliceutils.ByImageSize(images))   // there will write to images

Please, show me the stacktrace or, better, explain how to reproduce the panic. It can help to find a common ground.

Removing your last commit and running test case 100 times in a row didn't produce any panics.

File: pkg/kubelet/container/testing/fake_runtime.go [hunk 2]
```diff
@@ -81,6 +81,12 @@ type FakeVersion struct {
 	Version string
 }
 
+func snapshot(imageList []kubecontainer.Image) []kubecontainer.Image {
```
Comment (haircommander) [inline pkg/kubelet/container/testing/fake_runtime.go:84]: nit: it's a style preference of mine to define functions below the calling function so the file can be read top to bottom. would you mind moving this below?

Comment (haircommander) [reply pkg/kubelet/container/testing/fake_runtime.go:84]: (that is, if we keep snapshot)

**Comment by haircommander** [comment]
I agree with @bart0sh I think it's more common to use a lock in cases like this

**Comment by lixd** [comment]
@bart0sh ,emm, i test `go test -run ./...` it's ok, but use `go test -run ./... -race` waring a data race,stack :
```go
==================
WARNING: DATA RACE
Read at 0x00c0004f3c80 by goroutine 106:
  k8s.io/kubernetes/pkg/kubelet/container/testing.(*FakeRuntime).GetImageRef()
      /Users/lixueduan/17x/github/fork/kubernetes/pkg/kubelet/container/testing/fake_runtime.go:358 +0x208
  k8s.io/kubernetes/pkg/kubelet/images.(*realImageGCManager).detectImages()
      /Users/lixueduan/17x/github/fork/kubernetes/pkg/kubelet/images/image_gc_manager.go:216 +0x1ac
  k8s.io/kubernetes/pkg/kubelet/images.(*realImageGCManager).Start.func1()
      /Users/lixueduan/17x/github/fork/kubernetes/pkg/kubelet/images/image_gc_manager.go:189 +0x58
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /Users/lixueduan/17x/github/fork/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x48
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /Users/lixueduan/17x/github/fork/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0x94
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /Users/lixueduan/17x/github/fork/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x110
  k8s.io/apimachinery/pkg/util/wait.Until()
      /Users/lixueduan/17x/github/fork/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161 +0x40
  k8s.io/kubernetes/pkg/kubelet/images.(*realImageGCManager).Start.func3()
      /Users/lixueduan/17x/github/fork/kubernetes/pkg/kubelet/images/image_gc_manager.go:188 +0x4c

Previous write at 0x00c0004f3c80 by goroutine 107:
  k8s.io/kubernetes/pkg/kubelet/util/sliceutils.ByImageSize.Swap()
      /Users/lixueduan/17x/github/fork/kubernetes/pkg/kubelet/util/sliceutils/sliceutils.go:51 +0x130
  k8s.io/kubernetes/pkg/kubelet/util/sliceutils.(*ByImageSize).Swap()
      <autogenerated>:1 +0x24
  sort.insertionSort()
      /usr/local/go/src/sort/zsortinterface.go:13 +0x88
  sort.pdqsort()
      /usr/local/go/src/sort/zsortinterface.go:73 +0x2c8
  sort.Sort()
      /usr/local/go/src/sort/sort.go:48 +0x6c
  k8s.io/kubernetes/pkg/kubelet/images.(*imageCache).set()
      /Users/lixueduan/17x/github/fork/kubernetes/pkg/kubelet/images/image_gc_manager.go:130 +0x90
  k8s.io/kubernetes/pkg/kubelet/images.(*realImageGCManager).Start.func2()
      /Users/lixueduan/17x/github/fork/kubernetes/pkg/kubelet/images/image_gc_manager.go:201 +0x1c8
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /Users/lixueduan/17x/github/fork/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x48
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /Users/lixueduan/17x/github/fork/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0x94
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /Users/lixueduan/17x/github/fork/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x110
  k8s.io/apimachinery/pkg/util/wait.Until()
      /Users/lixueduan/17x/github/fork/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161 +0x40
  k8s.io/kubernetes/pkg/kubelet/images.(*realImageGCManager).Start.func4()
      /Users/lixueduan/17x/github/fork/kubernetes/pkg/kubelet/images/image_gc_manager.go:196 +0x4c

Goroutine 106 (running) created at:
  k8s.io/kubernetes/pkg/kubelet/images.(*realImageGCManager).Start()
      /Users/lixueduan/17x/github/fork/kubernetes/pkg/kubelet/images/image_gc_manager.go:188 +0x1a4
  k8s.io/kubernetes/pkg/kubelet/images.TestDeleteUnusedImagesLimitByImageLiveTime()
      /Users/lixueduan/17x/github/fork/kubernetes/pkg/kubelet/images/image_gc_manager_test.go:462 +0x65c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x188
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x40

Goroutine 107 (running) created at:
  k8s.io/kubernetes/pkg/kubelet/images.(*realImageGCManager).Start()
      /Users/lixueduan/17x/github/fork/kubernetes/pkg/kubelet/images/image_gc_manager.go:196 +0x304
  k8s.io/kubernetes/pkg/kubelet/images.TestDeleteUnusedImagesLimitByImageLiveTime()
      /Users/lixueduan/17x/github/fork/kubernetes/pkg/kubelet/images/image_gc_manager_test.go:462 +0x65c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x188
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x40
==================

```
there is a data race,but not occured panic,the root case as i say FakeRuntime shared a variable to caller,so there maybe write and read at the same time.
Even if used lock in `GetImageRef` and `ListImages`,but when the caller get the result of `ListImages`, caller can write it without lock,this operate is out the range of `FakeRuntime`.
I still think we can use snapshot to fix this,if use lock we need coupling the manager and runtime.
```go
func (f *FakeRuntime) GetImageRef(_ context.Context, image kubecontainer.ImageSpec) (string, error) {
	f.Lock()
	defer f.Unlock()

	f.CalledFunctions = append(f.CalledFunctions, "GetImageRef")
	for _, i := range f.ImageList {
		if i.ID == image.Image {
			return i.ID, nil
		}
	}
	return "", f.InspectErr
}

func (f *FakeRuntime) ListImages(_ context.Context) ([]kubecontainer.Image, error) {
	f.Lock()
	defer f.Unlock()

	f.CalledFunctions = append(f.CalledFunctions, "ListImages")
	return f.ImageList, f.Err
}
```

File: pkg/kubelet/container/testing/fake_runtime.go (hunk 2 above)
Comment (lixd) [reply pkg/kubelet/container/testing/fake_runtime.go:84]: thanks,i weill fix this,if we keep snapshot, :)

**Comment by bart0sh** [comment]
FWIW, this change fixed the race for me:
```
diff --git a/pkg/kubelet/images/image_gc_manager.go b/pkg/kubelet/images/image_gc_manager.go
index 63fde0f14cc..183b729fc02 100644
--- a/pkg/kubelet/images/image_gc_manager.go
+++ b/pkg/kubelet/images/image_gc_manager.go
@@ -212,6 +212,9 @@ func (im *realImageGCManager) GetImageList() ([]container.Image, error) {
 func (im *realImageGCManager) detectImages(ctx context.Context, detectTime time.Time) (sets.String, error) {
        imagesInUse := sets.NewString()
 
+       im.imageCache.Lock()
+       defer im.imageCache.Unlock()
+
        // Always consider the container runtime pod sandbox image in use
        imageRef, err := im.runtime.GetImageRef(ctx, container.ImageSpec{Image: im.sandboxImage})
        if err == nil && imageRef != "" {
```

I don't is it better than using snapshotting as it's a bit implicit. @haircommander WDYT?

**Comment by bart0sh** [comment]
Another idea if we still prefer to use locking is to make ImageList lockable and use the same list for manager image cache and fake manager, e.g. ```manager.imageCache = imageCache{images: fakeRuntime.ImageList}```

File: pkg/kubelet/container/testing/fake_runtime.go [hunk 3]
```diff
@@ -362,7 +362,14 @@ func (f *FakeRuntime) ListImages(_ context.Context) ([]kubecontainer.Image, erro
 	defer f.Unlock()
 
 	f.CalledFunctions = append(f.CalledFunctions, "ListImages")
-	return f.ImageList, f.Err
+	// use snapshot to return a copy of ImageList,avoid data race with caller
```
Comment (haircommander) [inline pkg/kubelet/container/testing/fake_runtime.go:365]: tiny nit: can you add a space between `ImageList,` and `avoid`

**Comment by haircommander** [comment]
/lgtm

thanks @lixd

**Comment by lixd** [comment]
@haircommander this pr rebase and resolved conflict with other commit,need a re lgtm,thanks

File: pkg/kubelet/container/testing/fake_runtime.go (hunk 2 above)
Comment (haircommander) [reply pkg/kubelet/container/testing/fake_runtime.go:84]: hm this seems to have sneaked up again

Comment (lixd) [reply pkg/kubelet/container/testing/fake_runtime.go]: oh,i am sorry,fixed

> hm this seems to have sneaked up again

oh,i am sorry,fixed

**Comment by lixd** [comment]
@dchen1107 hi, need approval,please have a look~

**Comment by hailkomputer** [comment]
Hello, bug triage team lead here. I want to check the status and remind that the the code freeze is starting [01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/24b27b03), as we want to ensure that each PR has a chance to be merged.
Are we still targeting the 1.29 release?

**Comment by dims** [comment]
/approve

mostly testing related, it looks good to me and has been worked on with input with a lot of reviews and has lgtm(s) as well.

**Pull request merged**